File: package/main/src/Array/sortingHelpers/rangeValidator.ts
export interface ValidatedSortRange {
  startIndex: number;
  endIndex: number;
  shouldSort: boolean;
}

/**
 * Validates and adjusts the start and end indices for an operation on an array.
 * Ensures indices are within the bounds of the array.
 *
 * @template T The type of elements in the array.
 * @param {T[]} array The array.
 * @param {number} startIndex The desired starting index.
 * @param {number} endIndex The desired ending index.
 * @returns {ValidatedSortRange} An object containing the validated start index, end index,
 *                               and a boolean indicating if the operation should proceed on the range.
 */
export const validateRange = <T>(
  // Renamed from validateSortRange for more general use
  array: T[],
  startIndex: number,
  endIndex: number,
): ValidatedSortRange => {
  const length = array.length;
  // Handle empty arrays early
  if (length === 0) {
    return { startIndex: 0, endIndex: -1, shouldSort: false };
  }

  const validatedStartIndex = Math.max(0, Math.min(startIndex, length - 1));
  const validatedEndIndex = Math.max(
    validatedStartIndex,
    Math.min(endIndex, length - 1),
  );

  return {
    startIndex: validatedStartIndex,
    endIndex: validatedEndIndex,
    shouldSort: validatedEndIndex >= validatedStartIndex,
  };
};

File: package/umt_rust/src/array/sorting_helpers/range_validator.rs
/// Result of validating sort range indices
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ValidatedSortRange {
    pub start_index: usize,
    pub end_index: usize,
    pub should_sort: bool,
}

/// Validates and adjusts the start and end indices for an operation on an array.
/// Ensures indices are within the bounds of the array.
///
/// # Arguments
///
/// * `array_len` - Length of the array
/// * `start_index` - The desired starting index
/// * `end_index` - The desired ending index
///
/// # Returns
///
/// A `ValidatedSortRange` containing the validated start index, end index,
/// and a boolean indicating if the operation should proceed on the range.
pub fn validate_range(
    array_len: usize,
    start_index: usize,
    end_index: usize,
) -> ValidatedSortRange {
    // Handle empty arrays early
    if array_len == 0 {
        return ValidatedSortRange {
            start_index: 0,
            end_index: 0,
            should_sort: false,
        };
    }

    let validated_start_index = start_index.min(array_len - 1);
    let validated_end_index = end_index.min(array_len - 1).max(validated_start_index);

    ValidatedSortRange {
        start_index: validated_start_index,
        end_index: validated_end_index,
        should_sort: validated_end_index >= validated_start_index,
    }
}
